subsys/disk: add stm32 sdmmc disk access driver

anthonybrandon opened the pull request:
Add a disk access driver for the stm32 sdmmc component. The driver is
based around the stm32 cube HAL and uses the blocking API.

Signed-off-by: Anthony Brandon <anthony@amarulasolutions.com>

EDIT: Based on https://github.com/zephyrproject-rtos/zephyr/pull/24787 for fat_fs sample rework

anthonybrandon committed:
disk: add stm32 sdmmc disk access driver

Add a disk access driver for the stm32 sdmmc component. The driver is
based around the stm32 cube HAL and uses the blocking API.

Signed-off-by: Anthony Brandon <anthony@amarulasolutions.com>

erwango reviewed (changes requested):
Thanks for providing this. I think this will interest many users.

Now, it is not possible to merge this change as is and will require some adaptations. Indeed, it is not possible to introduce as is a vendor specific subsystem, you need to provide support via a generic driver API.
Luckily, you can base your work on https://github.com/zephyrproject-rtos/zephyr/pull/13214 which was perfectly fine in that regard.

anthonybrandon commented:
@erwango Can you give me some more guidance on how you would like me to change this?
Do you mean I shouldn't use the STM32 Cube HAL? I see other drivers that use it however, so I thought it would be okay.

erwango commented:
> Do you mean I shouldn't use the STM32 Cube HAL? I see other drivers that use it however, so I thought it would be okay.

@anthonybrandon, you can, no problem with that. Issue here is that you're pushing a subsystem driver directly based on a stm32 peripheral. What I'd advice you to do is to implement your SDMMC disk driver based on the SDMM API proposed in #13214. Then STM32 HAL would be used in the SDMMC driver.

anthonybrandon commented:
@erwango , okay, that makes sense to me. Looking at the sdmmc API  proposed in #13214 it seems a bit more low level than the one exposed in the stm32 cube HAL. So I don't think I can make it fit.

erwango commented:
@anthonybrandon 

>  Looking at the sdmmc API proposed in #13214 it seems a bit more low level than the one exposed in the stm32 cube HAL. So I don't think I can make it fit.

Indeed. Then I see two options:
- Rebase and reuse sdmmc driver provided in #13214
- Propose an alternative SDMMC API, on top of your HAL based STM32 driver. And implement your disk access subsytem on top of it.

anthonybrandon commented:
@erwango thanks. I'll give both options a try and see what works out better for me.

anthonybrandon commented:
@erwango I've updated my change with a new API.
I've named it sdmmc_hal, since I intend it to be used for implementations based on whatever is provided by the vendor HAL (stm32 cube in this case, but should work for the esp hal as well).
If anyone ever implements a generic sdmmc driver, then this could be removed.
I looked into doing that, but I am not familiar enough with the low level sdmmc commands, and the references I looked at all were doing slightly different things, so I wasn't confident in that approach.

galak commented:
dev-review (Mar 26): For now just implement this directly to the disk_access.h api and remove the middle sdmmc layer for now.  We can look at a sdmmc api once we see there are multiple users.

anthonybrandon commented:
@galak Should I just revert to my original pull request in that case?

erwango commented:
@anthonybrandon 

> @galak Should I just revert to my original pull request in that case?
Or am I misunderstanding what you are asking?

You got it right. Following dev-review meeting, it was decided the initial approach is ok as there is no others users for now for SDMMC api. Sorry for the back and forth change.

anthonybrandon commented:
@erwango Sorry, been busy recently. I need to rebase and retest the original patch. I'll try to have it done soon.

erwango commented on subsys/disk/Kconfig:
```diff
@@ -131,4 +131,23 @@ config DISK_SDHC_VOLUME_NAME
+	bool "STM32 SDMMC driver"
+	depends on HAS_STM32CUBE
+	select USE_STM32_HAL_SD
+	select USE_STM32_LL_SDMMC
```
This should not be needed if you're using only the .h part of the ll api.
Can you check this is really needed (since I haven't seen use of ll .c function in the driver below) ?

anthonybrandon replied:
Without it I get link errors like:
```
modules/hal/stm32/stm32cube/stm32l4xx/drivers/src/stm32l4xx_hal_sd.c:947: undefined reference to `SDMMC_ConfigData'
```
Maybe USE_STM32_HAL_SD should depend on USE_STM32_LL_SDMMC?

erwango replied:
ok, let's keep it as is

erwango commented on subsys/disk/disk_access_stm32_sdmmc.c:
```diff
@@ -0,0 +1,401 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
Please put following line here:
```
#define DT_DRV_COMPAT st_stm32_sdmmc
```
Then use DT macros that don't need compatible argument

erwango commented on subsys/disk/Kconfig:
```diff
@@ -131,4 +131,23 @@ config DISK_SDHC_VOLUME_NAME
+	depends on HAS_STM32CUBE
+	select USE_STM32_HAL_SD
+	select USE_STM32_LL_SDMMC
+	help
```
Please add default value based on device tree node presence:
```
	default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_SDMMC))
```
With following defined upper in the same file:
```
DT_COMPAT_ST_STM32_SDMMC := st,stm32-sdmmc
```

erwango commented on subsys/disk/disk_access_stm32_sdmmc.c:
```diff
@@ -0,0 +1,401 @@
+	return err;
+}
+
+#if DT_HAS_NODE(DT_INST(0, st_stm32_sdmmc))
+#define INST_0 DT_INST(0, st_stm32_sdmmc)
```
This part should not be needed with the Kconfig tweak proposed above.

erwango reviewed (changes requested):
Few comments, otherwise it seems good, but I'm adding some people to the review to cross-check.

One request though: we need the driver to be compiled as part of sanitycheck/CI.
Can you propose a sample/test that could make use of it (along with a board config)?

erwango commented:
^^@ABOSTM

anthonybrandon commented:
@erwango thanks for the review. I will try to address all your comments.
Regarding the codeowners build failure, who should I put as codeowner?

erwango commented:
> Regarding the codeowners build failure, who should I put as codeowner?

Yourself :-)

anthonybrandon commented:
@erwango I think I've addressed all your previous comments.
Regarding the sanitycheck test, I've added one that only performs build.
I selected the stm32l496g_disco board which should have an sdmmc slot.
I don't have this board, so I can't really check, but it's only used for building so I don't think it should matter. The gpio's I used in the overlay are not correct. The stm32l496g_disco has some expansion gpios that are used and as far as I can tell those are not supported. The stm32l496g_disco also has no pwr-gpio. So maybe that one should be optional.

erwango commented on CODEOWNERS:
```diff
@@ -477,6 +478,7 @@
 /tests/net/lib/mqtt_packet/               @jukkar @tbursztyka
 /tests/net/lib/coap/                      @rveerama1
 /tests/net/socket/                        @jukkar @tbursztyka @pfalcon
+/tests/subsys/disk/stm32_sdmmc            @anthonybrandon
```
/tests/subsys/disk/stm32_sdmmc/

erwango commented on tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay:
```diff
@@ -0,0 +1,16 @@
+		cd-gpios = <&gpioa 15 GPIO_ACTIVE_LOW>;
+	};
+};
+
```
Remove empty line

erwango commented on tests/subsys/disk/stm32_sdmmc/prj.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_ZTEST=y
```
tests are expected to provide normalized test verdict.
This should rather be moved to the sample section: samples/subsys

erwango commented on tests/subsys/disk/stm32_sdmmc/src/main.c:
```diff
@@ -0,0 +1,15 @@
+#endif
+
+void test_main(void)
+{
```
Thanks for proposing this sample.
Though, can you provide an even quite basic example on how to exercise APIs?

anthonybrandon replied:
For my project I just use it through the file system API. Would it be fine to enable FAT and then in the sample just mount it and write something?
Or should I directly use the disk API?

erwango replied:
> Would it be fine to enable FAT and then in the sample just mount it and write something?

@anthonybrandon would you manage `samples/subsys/fs/fat_fs` compiling/running on that board ?

erwango reviewed (changes requested):
For board support, can you also check stm32f469i_disco?

anthonybrandon commented:
@erwango 
> For board support, can you also check stm32f469i_disco?
I looked into it. It seems the low level defines are a bit different, and probably the clock configuration as well. Some of it can be handled by simply using some ifdef to select between SDIO_ and SDMMC_, but the clock enable part would be a bit more complex to ensure a good frequency is selected. As far as I know zephyr does not (yet) handle the PLLSAI clocks for stm32. So the low level part is done in this driver instead. That will differ for various SoC.
I'm not sure what the right way forward is.

erwango commented:
> @erwango
> 
> > For board support, can you also check stm32f469i_disco?
> 
> I looked into it. It seems the low level defines are a bit different, and probably the clock configuration as well. Some of it can be handled by simply using some ifdef to select between SDIO_ and SDMMC_, but the clock enable part would be a bit more complex to ensure a good frequency is selected. As far as I know zephyr does not (yet) handle the PLLSAI clocks for stm32. So the low level part is done in this driver instead. That will differ for various SoC.
> I'm not sure what the right way forward is.

If not too much trouble, we can afford a few #ifdef to allow support on some SoCs, and configuration can be provided using Kconfig. If it allows to have a sample working on a board available upstream, this is worth the pain.

hjuul commented on dts/bindings/mmc/st,stm32-sdmmc.yaml:
```diff
@@ -0,0 +1,20 @@
+    pwr-gpios:
+        type: phandle-array
+        required: true
+        description: Power pin
```
What does `pwr-gpios` do? Is it something custom for a particular board? 
If so I suggest you keep it out of the sdmmc binding, I suppose most implementations won't use a power pin with SD cards.

anthonybrandon replied:
Yes, it is for my board. I can remove it from the driver.

erwango replied:
I think it is ok to make it optional: `required: false`

hjuul commented on tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay:
```diff
@@ -0,0 +1,16 @@
+		label = "sdmmc1";
+		clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00000400>;
+		pwr-gpios = <&gpioe 12 GPIO_ACTIVE_HIGH>;
+		cd-gpios = <&gpioa 15 GPIO_ACTIVE_LOW>;
```
I'm not familiar with stm32l496g_disco, but a brief look at the schematics suggest that `PA15` is not the correct value for `cd-gpios`.

anthonybrandon replied:
I think I mentioned this in one of my earlier comments, but as far as I can tell the gpio it's connected to is on some peripheral that is not currently supported. I just picked the stm32l496g_disco because board is similar to what I have and it has sdcard. I don't actually have that board. I could make the card detect pin optional and have the driver assume the sdcard is present if the gpio is not set.

hjuul commented:
> > @erwango
> > > For board support, can you also check stm32f469i_disco?
> > 
> > 
> > I looked into it. It seems the low level defines are a bit different, and probably the clock configuration as well. Some of it can be handled by simply using some ifdef to select between SDIO_ and SDMMC_, but the clock enable part would be a bit more complex to ensure a good frequency is selected. As far as I know zephyr does not (yet) handle the PLLSAI clocks for stm32. So the low level part is done in this driver instead. That will differ for various SoC.
> > I'm not sure what the right way forward is.
> 
> If not too much trouble, we can afford a few #ifdef to allow support on some SoCs, and configuration can be provided using Kconfig. If it allows to have a sample working on a board available upstream, this is worth the pain.

@anthonybrandon is it at all possible to use STM32Cube HAL instead of LL when configuring clocks? That should make the code compatible across stm32 families. I'm trying to get this to work for stm32f746g_disco and can help verify the implementation on that platform.

anthonybrandon commented:
> 
> @anthonybrandon is it at all possible to use STM32Cube HAL instead of LL when configuring clocks? That should make the code compatible across stm32 families. I'm trying to get this to work for stm32f746g_disco and can help verify the implementation on that platform.

I'm not sure. I have to take a close look at what is available in the HAL. Looking at projects generated by the stm32cube tools the clock setup is done once at the start using `HAL_RCC_ClockConfig` which also sets the clocks for other peripherals. I can take a look and see if that is usable for setting the clock for just one peripheral.

anthonybrandon commented:
@hjuul As far as I can tell the HAL is different for L4 and F4 (and possibly other) SoCs specifically around the issue of sdmmc/sdio and the clock tree.
So even if it's possible to use HAL instead of LL there will need to be ifdefs.

hjuul commented:
> @hjuul As far as I can tell the HAL is different for L4 and F4 (and possibly other) SoCs specifically around the issue of sdmmc/sdio and the clock tree.
> So even if it's possible to use HAL instead of LL there will need to be ifdefs.

Ok, for stm32f746g_disco it works if I just remove the LL calls in `stm32_sdmmc_clock_enable()` (currently lines 43-59). This is because it uses the main PLL (not PLLSAI) as input to PLL48CLK by default.
So #ifdefs will work, the challenge is to use the right conditions (could be Kconfig).

hjuul commented on tests/subsys/disk/stm32_sdmmc/boards/stm32l496g_disco.overlay:
```diff
@@ -0,0 +1,16 @@
+ */
+
+/ {
+	sdmmc1 {
```
Should we also add sdmmc1 entries to the `.dtsi` files? That would make it easy to add support for sdmmc1 in board `.dts` files (don't need overlay files).

anthonybrandon replied:
Yes, if it's tested on the board I think it would probably be right to add support in the board dts instead of overlay.

erwango replied:
+1 to move things to .dtsi/.dts when possible.
A node should be defined in .dtsi (undet dts/arm/st/...), then edited in board .dts for customization.
Overlay files should be used for application specific purpose when needed

hjuul commented:
@anthonybrandon I've managed to get your code working on stm32f746_disco board. I mostly had to change board- and family specific files. If it is ok with you I'd like to add support for this board as part of this PR.
What is the best way for me to contribute the code? Can I add to this PR or should I send you a patch?

On another matter: Have you considered using the SDMMC with DMA? I will probably add that for my project, but perhaps that's a topic for a different  PR..

anthonybrandon commented:
> @anthonybrandon I've managed to get your code working on stm32f746_disco board. I mostly had to change board- and family specific files. If it is ok with you I'd like to add support for this board as part of this PR.
> What is the best way for me to contribute the code? Can I add to this PR or should I send you a patch?
> 

I'm not very familiar with githubs workflow. If there is a way for you to add to this PR I don't mind. Otherwise send me the patch and I will add it to this PR. If you upload your change to your own github I can pull it from there.

> On another matter: Have you considered using the SDMMC with DMA? I will probably add that for my project, but perhaps that's a topic for a different PR..

I considered it, but I think that is something for a separate PR.

erwango commented:
@hjuul Thanks for your help, this will help converging to a upstream ready solution.
A suggestion for contributing your code is to push a new PR that includes commits from this PR, so we can see the additions that are needed to work on stm32f746g_disco.
Btw, is there on board SD/MMC on this board?

anthonybrandon committed:
boards: arm: stm32l496g_disco: enable sdmmc support

Update the dtsi for stm32l471 (which the higher SoCs are based on) to
support the stm32-sdmmc disk access device. Enable the device for the
stm32l496g_disco board, and update the pinmuxing.
Note that the stm32l496g_disco board also has a card detect gpio
(MFX_GPIO8), but this is not supported yet. When not specified the
driver will assume a card is present.

Signed-off-by: Anthony Brandon <anthony@amarulasolutions.com>

anthonybrandon commented:
@erwango @hjuul 

- I've made both cd-gpios and pwr-gpios optional. If no card detect is specified it assumes there is a card. The reason I did this is becuase the stm32l496g_disco board has a card detect pin on an expansion device which is not supported as far as I can tell. 
- I've removed the test specific to this driver.
- I've updated the stm32l471.dtsi to add the device node. I've updated the stm32l496 board to enable it and have correct pinmuxing. 
- I've updated the fat_fs sample to compile for this board.

I'm looking forward to Hjuul's patch for stm32f7.

hjuul committed:
boards: arm: stm32f746g_disco: enable sdmmc support

Update dtsi and pinmux macros for stm32f7 family. Add sdmmc1 to dts file
for stm32f746g_disco. Also add board specific configuration file for
fat_fs sample.

Signed-off-by: Helge Juul <helge@fastmail.com>

hjuul commented:
> @hjuul Thanks for your help, this will help converging to a upstream ready solution.
> A suggestion for contributing your code is to push a new PR that includes commits from this PR, so we can see the additions that are needed to work on stm32f746g_disco.

@erwango  I've forked Zephyr, merged in the commits from this PR, committed my changes and pushed to https://github.com/hjuul/zephyr/tree/stm32f746g_disco-sdmmc.
Should I create a PR to `anthonybrandon:stm32-sdmmc` or to `zephyr:master` ?

> Btw, is there on board SD/MMC on this board?

Yes

anthonybrandon commented:
> @erwango I've forked Zephyr, merged in the commits from this PR, committed my changes and pushed to https://github.com/hjuul/zephyr/tree/stm32f746g_disco-sdmmc.
> Should I create a PR to `anthonybrandon:stm32-sdmmc` or to `zephyr:master` ?
I'll pick it from there and add it here.

anthonybrandon commented:
@hjuul I've updated the my patch series. Please confirm that it still works for you.
The base address is now taken from the device tree instead of the HAL macro SDMMC1, which does not exist for every SoC series and is more consistent with how other devices are handled in the dts.
I also made a small change to the pinmux in your patch to use the DT_HAS_NODE macro.

hjuul commented on dts/arm/st/l4/stm32l471.dtsi:
```diff
@@ -188,5 +188,13 @@
+			reg = <0x40012800 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00000400>;
+			status = "disabled";
+			label = "SDMMC1";
```
Minor detail: It seems everywhere else there is an underscore before the 1. So `SDMMC1` -> `SDMMC_1`.

anthonybrandon replied:
Fixed, thanks.

hjuul commented:
> @hjuul I've updated the my patch series. Please confirm that it still works for you.


I just checked with your latest changes, it still works for stm32f746g_disco.

> I also made a small change to the pinmux in your patch to use the DT_HAS_NODE macro.

Yeah, I chose to use Kconfig ifdef instead. Honestly I don't know what's the best choice here. @erwango , any opinion?

erwango commented:
> Yeah, I chose to use Kconfig ifdef instead. Honestly I don't know what's the best choice here. @erwango , any opinion?

We're moving toward using DT api whenever is possible,so DT_HAS_NODE is the good choice here.
Seems there are been good progress globally. I'll review in detail on monday.

erwango commented on subsys/disk/disk_access_stm32_sdmmc.c:
```diff
@@ -0,0 +1,398 @@
+	struct device *dev = disk->dev;
+	struct stm32_sdmmc_priv *priv = dev->driver_data;
+
+	LOG_DBG("%s", __func__);
```
Can you check that LOG_DBG is not already returning function name ?

anthonybrandon replied:
I think it does. Should I just remove all the DBG messages?

erwango commented on samples/subsys/fs/fat_fs/prj_stm32f746g_disco.conf:
```diff
@@ -0,0 +1,5 @@
+CONFIG_DISK_ACCESS=y
+CONFIG_LOG=y
+CONFIG_FILE_SYSTEM=y
+CONFIG_FAT_FILESYSTEM_ELM=y
+CONFIG_PRINTK=y
```
IIC These are all already available in prj.conf, so this file should'nt be needed

anthonybrandon replied:
The default prj.conf for that sample has:
`CONFIG_DISK_ACCESS_SPI_SDHC=y`
Which causes a build failure because the stm32 boards don't have that.

erwango replied:
As I commented in #24787, CONFIG_DISK_ACCESS_SPI_SDHC=y might be better put in each boards .conf so you can have this board to be treated like other boards under /boards

erwango replied:
@anthonybrandon https://github.com/zephyrproject-rtos/zephyr/pull/24787 is now taking this into account.
You just need to rebase and use same config scheme

anthonybrandon replied:
okay, done.

erwango replied:
@anthonybrandon This is not exactly what I expect. Sorry if not clear. The following are already available in prj.cnf:
```
CONFIG_DISK_ACCESS=y
CONFIG_LOG=y
CONFIG_FILE_SYSTEM=y
CONFIG_FAT_FILESYSTEM_ELM=y
```
So, if you really need printk, you just need to add a file `samples/subsys/fs/fat_fs/boards/stm32f746g_disco.conf` with the following:
```
CONFIG_PRINTK=y
```

erwango commented on samples/subsys/fs/fat_fs/prj_stm32l496g_disco.conf:
```diff
@@ -0,0 +1,5 @@
+CONFIG_DISK_ACCESS=y
+CONFIG_LOG=y
+CONFIG_FILE_SYSTEM=y
+CONFIG_FAT_FILESYSTEM_ELM=y
+CONFIG_PRINTK=y
```
IIC These are all already available in prj.conf, so this file should'nt be needed

MaureenHelm replied:
This conf file isn't under `boards/`, so it gets used instead of `prj.conf`. I proposed a refactoring of this sample's conf files in #24787

erwango replied:
This seems cleaner indeed, txs @MaureenHelm.
@anthonybrandon can you use https://github.com/zephyrproject-rtos/zephyr/pull/24787/commits/3649c16046c083e2f6d3951791cc4916180901d4 and do you changes on top?

erwango reviewed (changes requested):
LGTM, 2 requests though:
-I think you can squash the 3 first commits.
-Also, can you update sample's README ?

anthonybrandon commented:
@erwango I've squashed the three first patches, and removed the debug prints. I've added a separate commit to update the fat_fs sample readme.

MaureenHelm commented on samples/subsys/fs/fat_fs/sample.yaml:
```diff
@@ -4,4 +4,5 @@ tests:
   sample.filesystem.fat_fs:
     build_only: true
     platform_whitelist: nrf52840_blip olimexino_stm32 mimxrt1050_evk
+        stm32l496g_disco
```
I proposed a refactoring of the sample yaml in #24787 to reduce the platform whitelisting. If merged, it means you would instead add `sdhc` to the supported list in your board yaml.

anthonybrandon replied:
Does that mean I should move `DISK_ACCESS_STM32_SDMMC` under `DISK_ACCESS_SDHC` in the Kconfig?

MaureenHelm replied:
That's a good point, I'm not sure if it makes sense to move it. Keeping it there means we'll either need `samples/subsys/fs/fat_fs/prj_stm32l496g_disco.conf` like you originally proposed, or move `CONFIG_DISK_ACCESS_SDHC=y` to board conf overlay files in `samples/subsys/fs/fat_fs/boards/`. I'm ok with either option.

@erwango what do you think?

anthonybrandon replied:
@MaureenHelm I checked and building with `CONFIG_DISK_ACCESS_SDHC=y` does work for the STM32 boards, since all it seems to do is enable the submenu in Kconfig. So it could just stay as is. It's just superfluous for the stm32_sdmmc driver.
Not having the SDHC_SPI enabled is enough to allow it to build. So from that standpoint it would be okay to leave as is.

erwango replied:
Yes, let's drop the readme.
For the DISK_ACCESS_SDHC, I'm not so sure, I don't think you need it enabled in your case, so it would be part of each board configuration.

anthonybrandon committed:
boards: stm32l496g_disco: Update board doc for sdmmc

Add entries to features and pinmux tables to reflect support for sdmmc.

Signed-off-by: Anthony Brandon <anthony@amarulasolutions.com>

anthonybrandon committed:
boards: stm32f746g_disco: Update board doc for sdmmc

Add entries to features and pinmux tables to reflect support for sdmmc.

Signed-off-by: Anthony Brandon <anthony@amarulasolutions.com>

hjuul commented:
@anthonybrandon I'm very eager to get these changes into master branch before the merge window close for release 2.3, which is Friday this week (May 8th).
Please let me know if there is anything I can do to help you speed things up.

anthonybrandon commented:
@hjuul other than maybe retesting that your board still works (I didn't change anything major so I don't see why it wouldn't) I don't think there is anything left to do at the moment.

hjuul commented:
> 
> 
> @hjuul other than maybe retesting that your board still works (I didn't change anything major so I don't see why it wouldn't) I don't think there is anything left to do at the moment.

Great. I just checked again with my board (stm32f746g_disco) using the latest commit in this PR, and it still works.
Thank you very much for providing this work @anthonybrandon, I really appreciate it!

@erwango , do you think it is realistic to get this PR into master before the feature merge window this friday?

erwango reviewed (changes requested):
@anthonybrandon not sure if this is an oversight or not, but stm32f746g_disco commit still requires a change

anthonybrandon commented:
@erwango yes, I thought I had removed that, but apparently I forgot.

erwango commented:
> @erwango , do you think it is realistic to get this PR into master before the feature merge window this friday?

@hjuul Hopefully yes. Btw, if you're satifsfied, approving the PR & mention in the review that this is working on your side will help

hjuul reviewed (approved):
I have verified that this is running fine on the stm32f746_disco board.

erwango commented on boards/arm/stm32l496g_disco/pinmux.c:
```diff
@@ -39,6 +39,14 @@ static const struct pin_config pinconf[] = {
 #ifdef CONFIG_PWM_STM32_2
 	{STM32_PIN_PA0, STM32L4X_PINMUX_FUNC_PA0_PWM2_CH1},
 #endif /* CONFIG_PWM_STM32_2 */
+#if DT_HAS_NODE(DT_NODELABEL(sdmmc1))
```
Replace `DT_HAS_NODE` with `DT_HAS_NODE_STATUS_OKAY`

anthonybrandon replied:
Done

erwango commented on boards/arm/stm32f746g_disco/pinmux.c:
```diff
@@ -58,6 +58,15 @@ static const struct pin_config pinconf[] = {
 	{STM32_PIN_PA11, STM32F7_PINMUX_FUNC_PA11_OTG_FS_DM},
 	{STM32_PIN_PA12, STM32F7_PINMUX_FUNC_PA12_OTG_FS_DP},
 #endif	/* CONFIG_USB_DC_STM32 */
+#if DT_HAS_NODE(DT_NODELABEL(sdmmc1))
```
Replace `DT_HAS_NODE` with `DT_HAS_NODE_STATUS_OKAY`

erwango reviewed (changes requested):
@anthonybrandon can you rebase on master and apply following changes?

erwango reviewed (approved):
Txs!

erwango commented:
@anthonybrandon , since #24787 has been merged, you can try to rebase.
This is not mandatory, but with some luck it will also make CI pass.

erwango commented on subsys/disk/disk_access_stm32_sdmmc.c:
```diff
@@ -0,0 +1,420 @@
+	return err;
+}
+
+#if DT_HAS_DRV_INST(0)
```
Can you try #if DT_HAS_NODE_STATUS_OKAY(DT_DRV_INST(0))?

anthonybrandon replied:
Sorry. Didn't build it myself after rebasing.

erwango reviewed (changes requested):
DT macro to be updated.
@anthonybrandon we're in live DT API update, some more changes are needed.
Thanks for bearing with us.

galak commented on dts/bindings/mmc/st,stm32-sdmmc.yaml:
```diff
@@ -0,0 +1,22 @@
+    clocks:
+        required: true
+    label:
+        type: string
```
shouldn't need `type: string` here.

anthonybrandon replied:
I assume I should make a new pull request for that? Or can I still update this one after it's merged?

anthonybrandon replied:
Fixed in #25132

carlescufi commented:
Merging pending small tweak to the binding, which can be provided later.
